File: server/voiceInterfaceService.ts
import OpenAI from 'openai';
import fs from 'fs/promises';
import path from 'path';
import type { Storage } from './storage';

function getOpenAIClient(): OpenAI | null {
  const baaSignedEnv = process.env.OPENAI_BAA_SIGNED;
  const isBaaSigned = baaSignedEnv === 'true' || baaSignedEnv === '1';
  
  if (!isBaaSigned || !process.env.OPENAI_API_KEY) {
    return null;
  }
  
  return new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
}

interface VoiceTranscription {
  text: string;
  duration: number;
  language: string;
}

interface VoiceResponse {
  text: string;
  audioUrl?: string;
}

class VoiceInterfaceService {
  private storage: Storage;

  constructor(storage: Storage) {
    this.storage = storage;
  }

  async transcribeAudio(audioFilePath: string): Promise<VoiceTranscription> {
    const openai = getOpenAIClient();

    if (!openai) {
      throw new Error('OpenAI BAA not signed. Voice transcription unavailable.');
    }

    try {
      const { createReadStream } = await import('fs');
      const transcription = await openai.audio.transcriptions.create({
        file: createReadStream(audioFilePath) as any,
        model: 'whisper-1',
        language: 'en',
      });

      return {
        text: transcription.text,
        duration: 0,
        language: 'en',
      };
    } catch (error) {
      console.error('Error transcribing audio:', error);
      throw error;
    }
  }

  async generateSpeech(text: string, voice: 'alloy' | 'echo' | 'fable' | 'onyx' | 'nova' | 'shimmer' = 'nova'): Promise<Buffer> {
    const openai = getOpenAIClient();

    if (!openai) {
      throw new Error('OpenAI BAA not signed. Text-to-speech unavailable.');
    }

    try {
      const response = await openai.audio.speech.create({
        model: 'tts-1',
        voice,
        input: text,
      });

      const buffer = Buffer.from(await response.arrayBuffer());
      return buffer;
    } catch (error) {
      console.error('Error generating speech:', error);
      throw error;
    }
  }

  async processVoiceFollowup(
    patientId: string,
    audioFilePath: string
  ): Promise<{ transcription: VoiceTranscription; aiResponse: VoiceResponse }> {
    const transcription = await this.transcribeAudio(audioFilePath);
    
    const openai = getOpenAIClient();

    let aiResponseText = 'Thank you for your update. Your doctor will review it soon.';

    if (openai) {
      const completion = await openai.chat.completions.create({
        model: 'gpt-4o',
        messages: [
          {
            role: 'system',
            content:
              'You are Agent Clona, a compassionate AI health assistant. Provide empathetic, supportive responses to patient health updates. Ask relevant follow-up questions but never provide medical diagnoses or treatment advice.',
          },
          {
            role: 'user',
            content: `Patient voice update: ${transcription.text}`,
          },
        ],
        temperature: 0.7,
        max_tokens: 200,
      });

      aiResponseText = completion.choices[0]?.message?.content || aiResponseText;
    }

    return {
      transcription,
      aiResponse: {
        text: aiResponseText,
      },
    };
  }
}

export let voiceInterfaceService: VoiceInterfaceService;

export function initVoiceInterfaceService(storage: Storage) {
  voiceInterfaceService = new VoiceInterfaceService(storage);
}

File: app/services/voice_interface_service.py
from typing import Dict
import openai
from app.config import settings, check_openai_baa_compliance


class VoiceInterfaceService:
    def __init__(self):
        self.openai_enabled = check_openai_baa_compliance()
        if self.openai_enabled and settings.OPENAI_API_KEY:
            openai.api_key = settings.OPENAI_API_KEY
    
    async def transcribe_audio(self, audio_file_path: str) -> Dict:
        if not self.openai_enabled:
            raise Exception("OpenAI BAA not signed. Voice transcription unavailable.")
        
        try:
            with open(audio_file_path, "rb") as audio_file:
                transcription = await openai.Audio.atranscribe(
                    model="whisper-1",
                    file=audio_file,
                    language="en"
                )
            
            return {
                "text": transcription.text,
                "duration": 0,
                "language": "en"
            }
        except Exception as e:
            print(f"Error transcribing audio: {e}")
            raise
    
    async def generate_speech(self, text: str, voice: str = "nova") -> bytes:
        if not self.openai_enabled:
            raise Exception("OpenAI BAA not signed. Speech synthesis unavailable.")
        
        try:
            response = await openai.Audio.speech.acreate(
                model="tts-1",
                voice=voice,
                input=text
            )
            
            return response.content
        except Exception as e:
            print(f"Error generating speech: {e}")
            raise
    
    async def process_voice_followup(self, patient_id: str, audio_file_path: str) -> Dict:
        transcription = await self.transcribe_audio(audio_file_path)
        
        return {
            "patient_id": patient_id,
            "transcription": transcription["text"],
            "analysis": "Voice followup processed successfully",
            "success": True
        }
